### PR TITLE
Introducing interface into LogListRefresher.

### PR DESCRIPTION
**Pull request by Mercurrent**
For testability of dependent services.

### CODE REVIEW

File: submission/loglist_refresher.go [hunk 1]
```diff
@@ -30,26 +30,29 @@ const (
...
-// observed.
-type LogListRefresher struct {
-	// updateMu limits LogListRefresher to a single Refresh() at a time.
+// LogListRefresher is interface for Log List updates watcher.
+type LogListRefresher interface {
+	Refresh() (*loglist.LogList, error)
+}
+
+// LogListRefresherImpl regularly reads Log-list and emits notifications when
+// updates/errors observed. Implements LogListRefresher interface.
```
Comment (daviddrysdale) [inline submission/loglist_refresher.go]: Seems odd to export a thing named `..Impl` -- callers shouldn't need to care about the implementation.  So maybe:
 - make this `logListRefresherImpl` (or just `logListRefresher`)
 - restore `NewLogListRefresher(string) LogListRefresher {}` (no need for `*` in return type now it's an interface) but have it internally return a `logListRefresherImpl` instance.

Comment (Mercurrent) [reply submission/loglist_refresher.go]: Renamed and changed signatures per suggestion.

**Comment by Mercurrent** [comment]
PTAL

File: submission/loglist_refresher_test.go [hunk 2]
```diff
@@ -8,7 +8,7 @@
 //
 // Unless required by applicable law or agreed to in writing, software
 // distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or ied.
```
Comment (daviddrysdale) [inline submission/loglist_refresher_test.go]: Accidental deletion here.

File: submission/loglist_refresher.go [hunk 3]
```diff
@@ -60,11 +63,13 @@ func (llr *LogListRefresher) Refresh() (*loglist.LogList, error) {
 	}
 
 	if bytes.Equal(json, llr.lastJSON) {
+		fmt.Printf("no changes")
```
Comment (daviddrysdale) [inline submission/loglist_refresher.go]: Probably want to drop these `Printf`s

Comment (Mercurrent) [reply submission/loglist_refresher.go]: Done

File: submission/loglist_refresher_test.go (hunk 2 above)
Comment (Mercurrent) [reply submission/loglist_refresher_test.go]: Done

**Comment by Mercurrent** [comment]
PTAL

File: submission/loglist_refresher_test.go [hunk 4]
```diff
@@ -184,7 +184,7 @@ func TestNewLogListRefresherUpdate(t *testing.T) {
 			} else if gotErr && !tc.errRegexp.MatchString(err.Error()) {
 				t.Fatalf("llr.Refresh() = (_, %q), want err to match regexp %q", err, tc.errRegexp)
 			}
-			if diff := cmp.Diff(ll, tc.wantLl); diff != "" {
+			if diff := cmp.Diff(tc.wantLl, ll); diff != "" {
 				t.Errorf("llr.Refresh(): diff -want +got\n%s", diff)
```
Comment (daviddrysdale) [inline submission/loglist_refresher_test.go]: I've seen this the other way around (`cmp.Diff(got,want)`, `"-got +want"`, which would mean leaving the arguments and changing the error string) but I don't know if there's a widespread convention...

Comment (Mercurrent) [reply submission/loglist_refresher_test.go]: Good question.
Docs for Diff(x, y) say:
'At the start of each line, a "-" prefix indicates an element removed from x, a "+" prefix to indicates an element added to y.'

So if we wanted something but didn't get it (-want in the comment claims), that should be part of x. So tc.wantLl should be x, not y.

**Pull request merged**